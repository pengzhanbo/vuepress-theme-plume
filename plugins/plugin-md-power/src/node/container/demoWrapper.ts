import type { Markdown } from 'vuepress/markdown'
import { resolveAttrs } from '.././utils/resolveAttrs.js'
import { createContainerPlugin } from './createContainer.js'

interface DemoWrapperAttrs {
  title?: string
  img?: string
  noPadding?: boolean
  height?: string
}

/**
 * :::demo-wrapper img no-padding title="xxx" height="100px"
 * :::
 */
export function demoWrapperPlugin(md: Markdown): void {
  createContainerPlugin(md, 'demo-wrapper', {
    before: (info: string) => {
      const { attrs } = resolveAttrs<DemoWrapperAttrs>(info)
      const wrapperClasses: string[] = ['demo-wrapper']
      let containerStyle = ''
      if (attrs.title)
        wrapperClasses.push('has-title')

      if (attrs.img)
        wrapperClasses.push('only-img')

      if (attrs.noPadding)
        wrapperClasses.push('no-padding')

      if (attrs.height) {
        const h = Number.parseFloat(attrs.height) === Number(attrs.height) ? `${attrs.height}px` : attrs.height
        containerStyle += `--demo-container-height: ${h};`
        wrapperClasses.push('has-height')
      }

      return `<div class="${wrapperClasses.join(' ')}">
  <div class="demo-head">
    <div class="demo-ctrl"><i></i><i></i><i></i></div>
    ${attrs.title ? `<h4 class="demo-title"><p>${attrs.title}</p></h4>` : ''}
  </div>
  <div class="demo-container" ${containerStyle ? `style="${containerStyle}"` : ''}>\n`
    },
    after: () => '</div></div>',
  })
}
