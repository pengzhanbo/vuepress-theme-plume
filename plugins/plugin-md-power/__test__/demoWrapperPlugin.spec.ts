import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { demoWrapperPlugin } from '../src/node/container/demoWrapper.js'

describe('demoWrapperPlugin', () => {
  const md = new MarkdownIt().use(demoWrapperPlugin)
  it('should work', () => {
    const code = `\
::: demo-wrapper
content
:::

::: demo-wrapper title="test"
content
:::

::: demo-wrapper no-padding img height="100px"

[xxx](/img.jpg)
:::

::: demo-wrapper no-padding img height="100"

[xxx](/img.jpg)
:::
`
    expect(md.render(code)).toMatchSnapshot()
  })
})
