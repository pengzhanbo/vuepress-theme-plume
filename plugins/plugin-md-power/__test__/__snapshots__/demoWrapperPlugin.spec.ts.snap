// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`demoWrapperPlugin > should work 1`] = `
"<div class="demo-wrapper">
  <div class="demo-head">
    <div class="demo-ctrl"><i></i><i></i><i></i></div>
    
  </div>
  <div class="demo-container" >
<p>content</p>
</div></div><div class="demo-wrapper has-title">
  <div class="demo-head">
    <div class="demo-ctrl"><i></i><i></i><i></i></div>
    <h4 class="demo-title"><p>test</p></h4>
  </div>
  <div class="demo-container" >
<p>content</p>
</div></div><div class="demo-wrapper only-img no-padding has-height">
  <div class="demo-head">
    <div class="demo-ctrl"><i></i><i></i><i></i></div>
    
  </div>
  <div class="demo-container" style="--demo-container-height: 100px;">
<p><a href="/img.jpg">xxx</a></p>
</div></div><div class="demo-wrapper only-img no-padding has-height">
  <div class="demo-head">
    <div class="demo-ctrl"><i></i><i></i><i></i></div>
    
  </div>
  <div class="demo-container" style="--demo-container-height: 100px;">
<p><a href="/img.jpg">xxx</a></p>
</div></div>"
`;
